src/signature.cpp

```diff
@@ -26,102 +26,51 @@ namespace mococrw {
 
 using namespace openssl;
 
-std::vector<uint8_t> SignatureUtils::create(AsymmetricPrivateKey &privateKey,
-                                            const RSAPadding &padding,
-                                            const std::string message)
-{
-    std::vector<uint8_t> messageDigest;
-
-    if(privateKey.getType() != AsymmetricKey::KeyTypes::RSA){
-        throw MoCOCrWException("Functionality is only supported for RSA keys");
-    }
-
-    try {
-        messageDigest = digestMessage(message, getHashing(padding));
-    }
-    catch (const OpenSSLException &e) {
-        throw MoCOCrWException(e.what());
-    }
+using SSL_SIGNATURE_DATA_Ptr = std::unique_ptr<unsigned char, SSLFree<unsigned char>>;
 
-    return create(privateKey, padding, messageDigest);
-}
-
-std::vector<uint8_t> SignatureUtils::create(AsymmetricPrivateKey &privateKey,
-                                            const RSAPadding &padding,
-                                            std::vector<uint8_t> messageDigest)
+std::vector<uint8_t> SignatureUtils::create(AsymmetricSignCtx &ctx,
+                                            const std::vector<uint8_t> &messageDigest)
 {
-    std::vector<uint8_t> signature;
-    const auto keyCtx = _EVP_PKEY_CTX_new(privateKey.internal());
-    std::unique_ptr<unsigned char, SSLFree<unsigned char>> sig;
-    size_t siglen;
-
-    if(privateKey.getType() != AsymmetricKey::KeyTypes::RSA){
-        throw MoCOCrWException("Functionality is only supported for RSA keys");
-    }
+    size_t sigLen;
+    SSL_SIGNATURE_DATA_Ptr signatureData;
 
     try {
-        if (!keyCtx.get()) {
-            throw MoCOCrWException("Context is empty");
-        }
+        auto keyCtx = _EVP_PKEY_CTX_new(ctx.getKey().internal());
         _EVP_PKEY_sign_init(keyCtx.get());
-        setUpContext(keyCtx.get(), padding);
+
+        setupOpenSSLCtx(ctx, keyCtx, RSAPadding::OperationTypes::Sign);
 
         // This determines the buffer length
         _EVP_PKEY_sign(keyCtx.get(),
                        nullptr,
-                       &siglen,
+                       &sigLen,
                        reinterpret_cast<const unsigned char *>(messageDigest.data()),
                        messageDigest.size());
 
-        _CRYPTO_malloc_init();
-        sig.reset(static_cast<unsigned char *>(_OPENSSL_malloc(siglen)));
+        signatureData.reset(static_cast<unsigned char *>(_OPENSSL_malloc(sigLen)));
```

Comment:
is there a reason why this memory needs to be managed by openssl? Otherwise - just use std::vector as buffer.